staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,1010 @@
+// Each device in a given pool, regardless of how many ResourceSlices, must have a unique name.
+// The ResourceSlice in which a device gets published may change over time. The unique identifier
+// for a device is the tuple <driver name>, <pool name>, <device name>.
+//
+// Whenever a driver needs to update a pool, it increments the pool.Spec.Pool.Generation number
+// and updates all ResourceSlices with that new number and new resource definitions. A consumer
+// must only use ResourceSlices with the highest generation number and ignore all others.
+//
+// When allocating all resources in a pool matching certain criteria or when
+// looking for the best solution among several different alternatives, a
+// consumer should check the number of ResourceSlices in a pool (included in
+// each ResourceSlice) to determine whether its view of a pool is complete and
+// if not, should wait until the driver has completed updating the pool.
+//
+// For resources that are not local to a node, the node name is not set. Instead,
+// the driver may use a node selector to specify where the devices are available.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
+type ResourceSlice struct {
+	metav1.TypeMeta `json:",inline"`
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// Contains the information published by the driver.
+	//
+	// Changing the spec increments the generation number.
+	Spec ResourceSliceSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+}
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
+type ResourceSliceSpec struct {
+	// Driver identifies the DRA driver providing the capacity information.
+	// A field selector can be used to list only ResourceSlice
+	// objects with a certain driver name.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// +required
+	Driver string `json:"driver" protobuf:"bytes,1,name=driver"`
+
+	// Pool describes the pool that this ResourceSlice belongs to.
+	//
+	// +required
+	Pool ResourcePool `json:"pool" protobuf:"bytes,2,name=pool"`
+
+	// NodeName identifies the node which provides the devices.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to ResourceSlices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// Exactly one of NodeName and NodeSelector must be set.
+	//
+	// +optional
+	NodeName string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// NodeSelector defines which nodes have access to the devices in the pool.
+	// If it is specified, but empty (has no terms), all nodes have access.
```

Comment:
Coming back to this: You are right that NodeSelector is usually additive.  The more terms you add, the more nodes are selected.  After looking at the other uses of NodeSelector, this does sort of stand out.  The problem is that we're trying to make "nothing" mean "something" again.

If you don't like this inconsistency then we could either add an explict `AllNodes bool` (which satisfies "exactly one") or add a discriminator as we did with CountMode.

```
slice:
    spec:
        availableFrom:
            mode: (Node | NodeSelector | AllNodes)
            nodeName: foo // one node
            nodeSelector: // a set of nodes
```

Reply:
I'm okay with keeping the `empty = all nodes have access`. This is set and consumed primarily by controllers, not humans, so I think it's okay to be slightly inconsistent with other usages.

Reply:
LOL, now you have me doubting.  How bad would `AllNodes bool` be?  My worry is that there is code which knows how to evaluate a NodeSelector, and this "empty" selector will evaluate to nothing.

Reply:
I'm okay with `AllNodes bool`.

Reply:
Done.